### ISSUE TITLE
[Coding style] Usage of Qed

### ISSUE THREAD

**Issue by nmvdw** [UniMath/UniMath#1416]
The following is written in the style guide (README.md):

> Do not end a proof with Qed., except with Goal, for that may prevent later computations.

However, this convention is not followed in UniMath, and there are valid reasons to deviate from it. The main reason is the efficiency of type checking. Several structures (categories, functors, bicategories, pseudofunctors) are equipped with laws that are never used for computation. If those would be transparent, then the amount of memory used will be higher than necessary, causing type checking to be slower. In addition, one might want to make certain proofs, such as lemmas involving equalities on natural numbers, opaque.

Note that opacity (be it with `Qed` or with `abstract`) should always be handled with care. For example, when defining an equivalence using `gradth`, one should avoid making the inverse opaque. That would block computation, and as a result, the resulting definition is more difficult to use.

For this reason, I propose to update the coding style to provide clear guidelines on when the usage of `Qed` and `abstract` is permissible and when not. This would also make it clearer to people when to use these features, and it would prevent the compilation from being unnecessarily long.

My suggestion would be something in the lines of

> Only use `Qed` for definitions/theorems/lemmas of which the proof should not never be unfolded. Do not make computationally relevant parts of definitions opaque.

**Comment by DanGrayson** [comment]
Somehow we should clarify how one might conclude that a proof should never be unfolded.  Also, one form of computation impeded by Qed is equality checking, which spoils proofs by reflexivity.  One thing that might make it safe to mark a proof opaque is if one can ensure that it's the only proof of that statement in the repository.

**Comment by nmvdw** [comment]
> Also, one form of computation impeded by Qed is equality checking, which spoils proofs by reflexivity.

Indeed. there is a trade-off between the efficiency of type checking and the convenience of computations. To deal with this situation, it is important that the type of something that is made opaque, should be a proposition. 

> One thing that might make it safe to mark a proof opaque is if one can ensure that it's the only proof of that statement in the repository.

This is a reasonable guideline and it is followed in the library. However, there is a corner case. For functors, it is often convenient to make the `is_functor`-part opaque. Using composition, we can generate multiple multiple proofs of functoriality as witnessed by the following fragment. 

```coq
Goal ∏ (C₁ C₂ : category) (F : C₁ ⟶ C₂), is_functor F.
Proof.
  intros C₁ C₂ F.
  exact (is_functor_composite F (functor_identity _)).
Qed.
```

I would say that this does not violate the principle.

**Comment by benediktahrens** [comment]
@nmvdw : thanks for raising this issue. Indeed, the current advice in the style guide is not adequate and needs to be changed. However, it seems difficult to give a general rule that is reasonably concise, does not quantify over all code like in what Dan is mentioning, and matches our experience.

In case we don't manage to formulate a rule, we should at least remove the current advice.

**Comment by DanGrayson** [comment]
Niels, I don't understand your remark about `is_functor`.  Could you clarify?

**Comment by nmvdw** [comment]
Let's look at the proof that the identity functor is a functor. In the library, only `is_functor_identity` is a proof of this result. However, we can generate more proofs of this fact by using that the composition of functors is again a functor. So, in a certain sense, there is more than 1 proof of this fact, and that would violate the guideline you gave. 

> However, it seems difficult to give a general rule that is reasonably concise, does not quantify over all code like in what Dan is mentioning, and matches our experience.

Indeed, but I think it would be good to give some guidelines. For example:
- If something is made opaque, then its type should be a proposition.
- It is often a good idea to split the definition of structures in data and properties. In concrete instances, the proof of the properties can be made opaque.
- In these cases, it is unnecessary to write multiple proofs of properties.

**Comment by DanGrayson** [comment]
> Let's look at the proof that the identity functor is a functor. In the library, only `is_functor_identity` is a proof of this result. However, we can generate more proofs of this fact by using that the composition of functors is again a functor. So, in a certain sense, there is more than 1 proof of this fact, and that would violate the guideline you gave.
> 

Right, but those other proofs don't have to occur in the repository.  Simply avoid formalizing them.

**Comment by DanGrayson** [comment]
> Indeed, but I think it would be good to give some guidelines. For example:
> 
> * If something is made opaque, then its type should be a proposition.
> * It is often a good idea to split the definition of structures in data and properties. In concrete instances, the proof of the properties can be made opaque.
> * In these cases, it is unnecessary to write multiple proofs of properties.


Maybe replace the third point with "It is undesirable to write multiple opaque proofs of properties, for then proofs of equality of objects containing them cannot be accomplished by reflexivity."

**Comment by benediktahrens** [comment]
This issue seems solved by #1428 , closing.
